dali/util/user_stream.h

```diff
@@ -36,48 +39,53 @@ class DLL_PUBLIC UserStream {
   /**
    * @brief Gets UserStream instance
    */
-  DLL_PUBLIC static UserStream* Get() {
+  DLL_PUBLIC static UserStream *Get() {
     static UserStream us;
     return &us;
   }
 
   /**
-   * @brief Obtains cudaStream_t for provided buffer. If there is no for given device,
+   * @brief Obtains cudaStream_t for provided Tensor. If there is no for given device,
+   * new one is created and stored in the internal map
    */
-  //  new one is created and stored in the internal map
-  DLL_PUBLIC cudaStream_t GetStream(const dali::Buffer<GPUBackend> &b) {
-    size_t dev = GetDeviceForBuffer(b);
-    std::lock_guard<std::mutex> lock(m_);
-    auto it = streams_.find(dev);
-    if (it != streams_.end()) {
-      return it->second;
-    } else {
-      DeviceGuard g(dev);
-      constexpr int kDefaultStreamPriority = 0;
-      CUDA_CALL(cudaStreamCreateWithPriority(&streams_[dev], cudaStreamNonBlocking,
-                                             kDefaultStreamPriority));
-      return streams_.at(dev);
-    }
+  DLL_PUBLIC cudaStream_t GetStream(const dali::Tensor<GPUBackend> &t) {
+    return GetStream(GetDeviceForBuffer(t));
   }
 
   /**
-   * @brief Synchronizes on the device where given buffer b exists
+   * @brief Obtains cudaStream_t for provided TensorList. If there is no for given device,
+   * new one is created and stored in the internal map
    */
-  DLL_PUBLIC void WaitForDevice(const dali::Buffer<GPUBackend> &b) {
-    size_t dev = GetDeviceForBuffer(b);
-    DeviceGuard g(dev);
-    CUDA_CALL(cudaDeviceSynchronize());
+  DLL_PUBLIC cudaStream_t GetStream(const dali::TensorList<GPUBackend> &tl) {
+    return GetStream(GetDeviceForBuffer(tl));
   }
 
   /**
-   * @brief Synchronizes on the the stream where buffer b was created
+   * @brief Synchronizes on the device where given Tensor t exists
    */
-  DLL_PUBLIC void Wait(const dali::Buffer<GPUBackend> &b) {
-    size_t dev = GetDeviceForBuffer(b);
-    DALI_ENFORCE(streams_.find(dev) != streams_.end(),
-        "Can only wait on user streams");
-    DeviceGuard g(dev);
-    CUDA_CALL(cudaStreamSynchronize(streams_[dev]));
+  DLL_PUBLIC void WaitForDevice(const dali::Tensor<GPUBackend> &t) {
```

Comment:
```suggestion
  template<template<GPUBackend>class BatchContainer>
  DLL_PUBLIC void WaitForDevice(const BatchContainer<GPUBackend> &t) {
```
?

Reply:
That's the other alternative, I just went with two overloads, I guess a coin flip to decide.

Reply:
Up to you